### ISSUE TITLE
Add support for a customizer function for deep observable construction

### ISSUE THREAD

**Issue by daedalus28** [mobxjs/mobx#1539]
I'd like to be able to provide a customizer function instead of just an object to the `observable` constructor. The goal is to opt out of specific things being observable by some arbitrary criteria instead of explicitly by key name. This comes in handy when you have a dynamic structure where you don't know the exact shape, but you do know _some_ things about it.

In case I've missed something that's already available, here's my specific use case:
I have a deeply nested dynamic tree structure. The majority of the values in it can and should be observable, but there are few cases where it might be wasteful. To make this more concrete, imagine a scenario like this:

```js
let tree = observable({
  id: 1,
  type: 'group'
  value: SomeSimpleObject,
  children: [{
    id: 2,
    type: 'group',
    value: SomeSimpleObject,
    children: [{
      id: 3
      type: 'result',
      value: SomeReallyBigBlobFromTheServer
    }]
  }]
})
```

`children` can be dynamically added and removed and their `value` might change, but the contents of `value` would only ever be completely replaced - so I'd want `value` to be shallow. At the start of my app, I won't have the full structure because child `2` and `3` might not be added until much later. My current plan is to manually create each child using observable so they can use decorators, but an api that would let me customize the construction of future items pushed into the children array based on their `type` string, for example, would be ideal and could live in one place.

**Comment by spion** [comment]
Would https://github.com/mobxjs/serializr help with this?

**Comment by daedalus28** [comment]
@spion It wouldn't because the issue isn't one-time deserialization, it's about mutations that occur over the lifetime of the object

**Comment by mweststrate** [comment]
I think there are two ways to address this, both being more clear (despite being more code) than what you are roughly propossing

1. use classes instead of plain objects. After all, you have specific knowledge and behavior for specific fields
2. preprocess items before adding them to the tree. This means that you indeed have to design your api in such a way that you have that kind of control.

Also, you might want to make sure that non-shallow observables have actually a measurable impact on performance, as the clarity of everything in your tree being deep observables might exceed the value of saving a few clock cycles. 

In other words, I think this problem can be quite well solved by using some utility methods to modify your observables rather than introducing yet another way to configure observables in the mobx api. There are too many options already :-).
